Unable to build ViSP with Visual Studio 2010

A lot of build issues occur during the build of `3rdparty/apriltag`.
The reason is that msvc 10 doesn't support C99 requested by apriltag.

Issue in `3rdparty/apriltag`

````
2>  apriltag.c
2>d:\marc\visp\3rdparty\apriltag\common/matd.h(58): warning C4200: nonstandard extension used : zero-sized array in struct/union
2>          Cannot generate copy-ctor or copy-assignment operator when UDT contains a zero-sized array
2>d:\marc\visp\3rdparty\apriltag\common\time_util.h(36): fatal error C1083: Cannot open include file: 'stdbool.h': No such file or directory
````

Issue in `modules\vision\src\pose-estimation\vpPoseRansac.cpp`

````
1>..\..\..\visp\modules\vision\src\pose-estimation\vpPoseRansac.cpp(523): error C2440: 'initializing' : cannot convert from 'std::_Tree_iterator<_Mytree>' to 'std::_Tree_const_iterator<_Mytree>'
1>          with
1>          [
1>              _Mytree=std::_Tree_val<std::_Tmap_traits<vpPoint,size_t,`anonymous-namespace'::CompareObjectPointDegenerate,std::allocator<std::pair<const vpPoint,size_t>>,false>>
1>          ]
1>          and
1>          [
1>              _Mytree=std::_Tree_val<std::_Tmap_traits<vpPoint,size_t,std::less<vpPoint>,std::allocator<std::pair<const vpPoint,size_t>>,false>>
1>          ]
1>          No constructor could take the source type, or constructor overload resolution was ambiguous
1>..\..\..\visp\modules\vision\src\pose-estimation\vpPoseRansac.cpp(524): error C2678: binary '!=' : no operator found which takes a left-hand operand of type 'std::_Tree_const_iterator<_Mytree>' (or there is no acceptable conversion)
1>          with
1>          [
1>              _Mytree=std::_Tree_val<std::_Tmap_traits<vpPoint,size_t,std::less<vpPoint>,std::allocator<std::pair<const vpPoint,size_t>>,false>>
1>          ]
1>          C:\Program Files\Microsoft Visual Studio 10.0\VC\include\system_error(425): could be 'bool std::operator !=(const std::error_code &,const std::error_condition &)' [found using argument-dependent lookup]
1>          C:\Program Files\Microsoft Visual Studio 10.0\VC\include\system_error(432): or       'bool std::operator !=(const std::error_condition &,const std::error_code &)' [found using argument-dependent lookup]
1>          D:\Marc\visp\modules\core\include\visp3/core/vpColor.h(222): or       'bool operator !=(const vpColor &,const vpColor &)'
1>          D:\Marc\visp\modules\core\include\visp3/core/vpImagePoint.h(309): or       'bool operator !=(const vpImagePoint &,const vpImagePoint &)'
1>          C:\Program Files\Microsoft SDKs\Windows\v7.0A\include\guiddef.h(197): or       'int operator !=(const GUID &,const GUID &)'
1>          C:\Program Files\Microsoft Visual Studio 10.0\VC\include\xtree(333): or       'bool std::_Tree_const_iterator<_Mytree>::operator !=(const std::_Tree_const_iterator<_Mytree> &) const'
1>          with
1>          [
1>              _Mytree=std::_Tree_val<std::_Tmap_traits<vpPoint,size_t,std::less<vpPoint>,std::allocator<std::pair<const vpPoint,size_t>>,false>>
1>          ]
1>          while trying to match the argument list '(std::_Tree_const_iterator<_Mytree>, std::_Tree_iterator<_Mytree>)'
1>          with
1>          [
1>              _Mytree=std::_Tree_val<std::_Tmap_traits<vpPoint,size_t,std::less<vpPoint>,std::allocator<std::pair<const vpPoint,size_t>>,false>>
1>          ]
1>          and
1>          [
1>              _Mytree=std::_Tree_val<std::_Tmap_traits<vpPoint,size_t,`anonymous-namespace'::CompareObjectPointDegenerate,std::allocator<std::pair<const vpPoint,size_t>>,false>>
1>          ]
````

Issue in `modules\core\test\math\testMatrixInverse.cpp`

````
162>..\..\..\visp\modules\core\test\math\testMatrixInverse.cpp(265): error C2668: 'sqrt' : ambiguous call to overloaded function
162>          C:\Program Files\Microsoft Visual Studio 10.0\VC\include\math.h(589): could be 'long double sqrt(long double)'
162>          C:\Program Files\Microsoft Visual Studio 10.0\VC\include\math.h(541): or       'float sqrt(float)'
162>          C:\Program Files\Microsoft Visual Studio 10.0\VC\include\math.h(127): or       'double sqrt(double)'
162>          while trying to match the argument list '(unsigned int)'
162>..\..\..\visp\modules\core\test\math\testMatrixInverse.cpp(266): error C2668: 'sqrt' : ambiguous call to overloaded function
162>          C:\Program Files\Microsoft Visual Studio 10.0\VC\include\math.h(589): could be 'long double sqrt(long double)'
162>          C:\Program Files\Microsoft Visual Studio 10.0\VC\include\math.h(541): or       'float sqrt(float)'
162>          C:\Program Files\Microsoft Visual Studio 10.0\VC\include\math.h(127): or       'double sqrt(double)'
162>          while trying to match the argument list '(unsigned int)'
````

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -425,7 +425,8 @@ VP_OPTION(USE_ZBAR        ZBAR         ""    "Include zbar support"         "" O
 VP_OPTION(USE_DMTX        DMTX         ""    "Include dmtx support"         "" ON IF NOT WINRT)
 VP_OPTION(USE_PCL         PCL          QUIET "Include Point Cloud Library support" "" ON IF (USE_REALSENSE OR USE_REALSENSE2) AND NOT WINRT AND NOT IOS)
 VP_OPTION(WITH_PTHREAD    ""           ""    "Build pthread as built-in library"   "" ON IF (NOT USE_PTHREAD) AND (WIN32 OR MINGW) AND (NOT WINRT_8_1) AND (NOT WINRT_8_0))
-VP_OPTION(WITH_APRILTAG   ""           ""    "Build AprilTag as built-in library"  "" ON IF (USE_PTHREAD OR WITH_PTHREAD) AND (NOT WINRT_8_1) AND (NOT WINRT_8_0))
+# Since C99 is not supported by MSVC 2010 or prior, we disable apriltag if MSVC < 2012
+VP_OPTION(WITH_APRILTAG   ""           ""    "Build AprilTag as built-in library"  "" ON IF (USE_PTHREAD OR WITH_PTHREAD) AND (NOT WINRT_8_1) AND (NOT WINRT_8_0) AND (NOT MSVC_VERSION LESS 1700))
 VP_OPTION(WITH_ATIDAQ     ""           ""    "Build atidaq-c as built-in library"  "" ON IF USE_COMEDI AND NOT WINRT)
 VP_OPTION(WITH_CLIPPER    ""           ""    "Build clipper as built-in library"   "" ON IF USE_OPENCV)
 VP_OPTION(WITH_LAPACK     ""           ""    "Build lapack as built-in library"    "" ON IF NOT USE_LAPACK)

--- a/ChangeLog
+++ b/ChangeLog
@@ -5,6 +5,13 @@
      Copyright (C) 2005 - 2017 by Inria. All rights reserved.
                       https://visp.inria.fr
 
+ViSP 3.2.0 (under development)
+  - New features
+  - Tutorials
+  - Bug fixed
+    . [#279] Unable to build ViSP with Visual Studio 2010
+
+----------------------------------------------
 ViSP 3.1.0 (released December 22nd, 2017)
   - Contributors:
     . This release was possible thanks to contributions from Fabien Spindler,

--- a/cmake/FindC99.cmake
+++ b/cmake/FindC99.cmake
@@ -73,7 +73,7 @@ if(C99_FLAG_DETECTED)
   " __compiler_doesnt_need_c99_flag)
 endif()
 
-if(NOT __compiler_doesnt_need_c99_flag)
+if(C99_FLAG_DETECTED AND NOT __compiler_doesnt_need_c99_flag)
   set(C99_C_FLAGS "${C99_C_FLAGS_INTERNAL}")
   set(C99_FOUND TRUE)
 else()

--- a/modules/core/test/math/testMatrixInverse.cpp
+++ b/modules/core/test/math/testMatrixInverse.cpp
@@ -262,8 +262,8 @@ int test_inverse(const std::vector<vpMatrix> &bench, const std::vector<vpMatrix>
 {
   for (unsigned int i = 0; i < bench.size(); i++) {
     vpMatrix I = bench[i] * result[i];
-    if (std::fabs(I.euclideanNorm() - sqrt(bench[0].AtA().getRows())) > 1e-10) {
-      std::cout << "Bad inverse[" << i << "]: " << I.euclideanNorm() << " " << sqrt(bench[0].AtA().getRows())
+    if (std::fabs(I.euclideanNorm() - sqrt((float)bench[0].AtA().getRows())) > 1e-10) {
+      std::cout << "Bad inverse[" << i << "]: " << I.euclideanNorm() << " " << sqrt((float)bench[0].AtA().getRows())
                 << std::endl;
       return EXIT_FAILURE;
     }

--- a/modules/core/test/math/testMatrixInverse.cpp
+++ b/modules/core/test/math/testMatrixInverse.cpp
@@ -262,8 +262,8 @@ int test_inverse(const std::vector<vpMatrix> &bench, const std::vector<vpMatrix>
 {
   for (unsigned int i = 0; i < bench.size(); i++) {
     vpMatrix I = bench[i] * result[i];
-    if (std::fabs(I.euclideanNorm() - sqrt((float)bench[0].AtA().getRows())) > 1e-10) {
-      std::cout << "Bad inverse[" << i << "]: " << I.euclideanNorm() << " " << sqrt((float)bench[0].AtA().getRows())
+    if (std::fabs(I.euclideanNorm() - sqrt((double)bench[0].AtA().getRows())) > 1e-10) {
+      std::cout << "Bad inverse[" << i << "]: " << I.euclideanNorm() << " " << sqrt((double)bench[0].AtA().getRows())
                 << std::endl;
       return EXIT_FAILURE;
     }

--- a/modules/vision/src/pose-estimation/vpPoseRansac.cpp
+++ b/modules/vision/src/pose-estimation/vpPoseRansac.cpp
@@ -520,7 +520,7 @@ bool vpPose::poseRansac(vpHomogeneousMatrix &cMo, bool (*func)(vpHomogeneousMatr
       }
 
       std::map<vpPoint, size_t, CompareImagePointDegenerate> filterImagePointMap;
-      for (std::map<vpPoint, size_t>::const_iterator it = filterObjectPointMap.begin();
+      for (std::map<vpPoint, size_t, CompareObjectPointDegenerate>::const_iterator it = filterObjectPointMap.begin();
            it != filterObjectPointMap.end(); ++it) {
         if (filterImagePointMap.find(it->first) == filterImagePointMap.end()) {
           filterImagePointMap[it->first] = it->second;

